src/onnx/parse_pad.cpp

```diff
@@ -0,0 +1,163 @@
+#include <migraphx/onnx/op_parser.hpp>
+#include <migraphx/onnx/checks.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/make_op.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace onnx {
+
+void calc_reflect_indices(std::vector<int>& indices, const int64_t num_dims)
+{
+    int k         = 0;
+    bool reversed = false;
+    // in reflect padding, if the num_pads > num_dims,
+    // compute the extra pad indices periodically, ex. ( 1, 2, 3, 2, 1, 0)
+    for(int& idx : indices)
+    {
+        if(k == num_dims - 1)
+            reversed = true;
+        if(k == 0)
+            reversed = false;
+        if(reversed)
+            k--;
+        else
+            k++;
+        idx = k;
+    }
+}
+
+instruction_ref reflect_pad(const onnx_parser::node_info& info,
+                            const std::vector<int64_t>& pads,
+                            instruction_ref input)
+{
+    size_t num_dims = pads.size() / 2;
+    std::vector<int> ldims(pads.begin(), pads.begin() + num_dims);
+    std::vector<int> rdims(pads.begin() + num_dims, pads.end());
+    assert(ldims.size() == rdims.size());
+
+    std::vector<int64_t> axes(num_dims);
+    std::iota(axes.begin(), axes.end(), int64_t{0});
+
+    // iterate over dimensions, starting from lowest dimension
+    for(int64_t i = num_dims - 1; i >= 0; i--)
+    {
+        auto axis   = i;
+        auto lcount = ldims.at(i);
+        auto rcount = rdims.at(i);
+        if(lcount == 0 and rcount == 0) // no padding for current dim
+            continue;
+
+        // calculate starts and ends for each iteration since shape may change
+        std::vector<size_t> dims = input->get_shape().lens();
+        std::vector<int64_t> starts(axes.size(), 0);
+        std::vector<int64_t> ends(dims.begin(), dims.end());
+        std::vector<instruction_ref> slices;
+
+        auto starts_it = starts.begin() + i;
+        auto ends_it   = ends.begin() + i;
+        auto dims_it   = dims.begin() + i;
+
+        std::vector<int> l_indices(lcount);
+        std::vector<int> r_indices(rcount);
+
+        // compute slice indices in a periodic fashion
+        calc_reflect_indices(l_indices, *dims_it);
+        calc_reflect_indices(r_indices, *dims_it);
+
+        for(int idx : l_indices)
+        {
+            *starts_it = idx;
+            *ends_it   = *starts_it + 1;
+            slices.push_back(info.mm->add_instruction(
```

Comment:
Can we just call info.add_instruction() here?

Reply:
Yea, done.